### ISSUE TITLE
Some RTF outlines with implicit hyphens considered invalid

### ISSUE THREAD

**Issue by sammdot** [benoit-pierre/plover_stroke#3]
From Jeff on the Plover Discord:
> The machine is translating everything as expected, but Plover is mistranslating (or not translating) a number of briefs. Most simply, the FPLT brief shows up as "-FPLT" in Word, rather than a period.
>
> ![Dictionary manager window showing various outlines for {.}, none of them with hyphens](https://user-images.githubusercontent.com/7952620/149609779-09117187-319b-4013-a1ca-adc02c52e712.png)

Aerick's response:
> I think what's happening here is Plover expects `-FPLT` but the period stroke is defined as `FPLT`
> so any entries that only use the right side keys will have to be reassigned

---

After reading the [RTF/CRE spec](https://web.archive.org/web/20201017075356/http://www.legalxml.org/workgroups/substantive/transcripts/cre-spec.htm), it seems to be much more lenient about including hyphens in outlines where the keys are already unambiguous:

> A hyphen is used in cases where there may be confusion between an initial and final consonant. For instance, to differentiate between the strokes `/RBGS` [Plover: `R-BGS`] and `/-RBGS` [Plover: `-RBGS`].

---

After some regex magic Jeff returned with not quite everything resolved:

> So my regex approach corrected about 6,000 entries, and things in Plover are way better now--but still not perfect, because this dictionary is also missing hyphens in many left+right strokes, like `THR-R` for "there are" ...nuts.
> ![An except from an RTF dictionary](https://user-images.githubusercontent.com/7952620/149609848-e21212dd-f47e-4cc9-84f2-1f781be24584.png)

In this case we see the same issue, in which both Rs are there and so the hyphen isn't strictly necessary, but the outline isn't getting parsed correctly.

We should be able to handle these cases as well, in order for RTF (and others) dictionaries from professionals and CAT systems to work without needing major changes. Plover's steno notation as implemented in this library is unambiguous, yes, but the input dictionaries with not-quite-canonical outlines should be considered valid as well and normalized into our notation.

**TL;DR:** `plover_stroke` is _too strict_ about hyphen usage, and some RTF/CRE dictionaries are failing to validate because of it.

**Comment by jrowberg** [comment]
FWIW, I used a slightly modified bit of code from **rabbitgrowth** on the Discord server to create a one-time fixer script:

https://pastebin.com/x7chWGDu

The ugly 4x substitution was because it was easier than trying to figure out how to make the regex engine match *either* the beginning/end of a line OR a forward slash; without catching all four cases (entire stroke, first stroke in a combo, middle stroke in a combo, or last stroke in a combo), the script missed some important cases. I'm not sure if something like this could be incorporated into the Plover parsing/matching code, but in case it can, have at it!

**Comment by benoit-pierre** [comment]
I'm confused: `plover_stroke` already parse all those strokes like that.

Here are the results of `normalize_steno`:
```
'R'       → ('R',)
'-R'      → ('-R',)
'FPLT'    → ('-FPLT',)
'THRPL'   → ('THR-PL',)
'THRPLT'  → ('THR-PLT',)
'THRPZ'   → ('THR-PZ',)
'THRR'    → ('THR-R',)
'THRRB'   → ('THR-RB',)
'THRRBD'  → ('THR-RBD',)
'THRRBGD' → ('THR-RBGD',)
'THRRBGS' → ('THR-RBGS',)
```

**Comment by jrowberg** [comment]
Okay, this is crazy. I have the latest Plover running (stock) from source on my machine, and I confirm what you said: if I enter "FPLT" alone as a new translation stroke in a new blank dictionary, Plover immediately adjusts the stroke to be "-FPLT" when I tab to the output field. There is no way that I can make "FPLT" remain (without a hyphen) in the stroke field.

Similarly, if I close Plover completely and manually edit my `test.json` file to remove the hyphen, then restart Plover, it normalizes it both in the dictionary view and in the live parser, and produces the correct output translation of `{.}`. This is all correct, and simultaneously not what I expected.

I am 100% certain that with the installation/import/work flow that my mom did:

1. Install Plover 4.0.0-dev10 binary distribution
2. Get 100k+ entry StenoCAT dictionary converted into RTF
3. Import converted RTF into Plover
4. Test with various strokes including right-hand-only FPLT

...we ended up with a not-normalized dictionary, as you can see from these screenshots posted in Discord a few days ago:

![image](https://user-images.githubusercontent.com/634312/150374364-6e8c08a6-a7c6-4ec1-9747-5cf4d2671aa5.png)

![image](https://user-images.githubusercontent.com/634312/150374434-bbb11067-1cba-4097-8854-99219bdc0e6e.png)

So it seems the `normalize_steno` function works perfectly, but there is some combination of import/load where a dictionary doesn't get normalized (assuming it isn't a bug that has been fixed between 4.0.0-dev10 and the current Github master commit).

Please ignore the fact that you can get a period with a zillion different strokes; she's in the process of cleaning up a very old dictionary mashed together from three different theories.

**Comment by benoit-pierre** [comment]
`4.0.0-dev10` does not use `plover_stroke`, and does not have the new RTF parser, so please try with the latest continuous release.
